core/src/main/java/org/elasticsearch/search/SearchModule.java

```diff
@@ -432,23 +432,73 @@ public AggregationSpec addResultReader(String writeableName, Writeable.Reader<?
 
     /**
      * Register a pipeline aggregation.
-     *
-     * @param reader reads the aggregation builder from a stream
-     * @param internalReader reads the {@link PipelineAggregator} from a stream
-     * @param bucketReader reads the {@link InternalAggregation} that represents a bucket in this aggregation from a stream
-     * @param aggregationParser reads the aggregation builder from XContent
-     * @param aggregationName names by which the aggregation may be parsed. The first name is special because it is the name that the reader
-     *        is registered under.
      */
-    public void registerPipelineAggregation(Writeable.Reader<? extends PipelineAggregationBuilder> reader,
-            Writeable.Reader<? extends PipelineAggregator> internalReader, Writeable.Reader<? extends InternalAggregation> bucketReader,
-            PipelineAggregator.Parser aggregationParser, ParseField aggregationName) {
+    public void registerPipelineAggregation(PipelineAggregationSpec spec) {
         if (false == transportClient) {
-            namedWriteableRegistry.register(PipelineAggregationBuilder.class, aggregationName.getPreferredName(), reader);
-            pipelineAggregationParserRegistry.register(aggregationParser, aggregationName);
+            namedWriteableRegistry.register(PipelineAggregationBuilder.class, spec.name.getPreferredName(), spec.builderReader);
+            pipelineAggregationParserRegistry.register(spec.parser, spec.name);
+        }
+        for (Map.Entry<String, Writeable.Reader<? extends PipelineAggregator>> resultReader : spec.resultReaders.entrySet()) {
+            namedWriteableRegistry.register(PipelineAggregator.class, resultReader.getKey(), resultReader.getValue());
+        }
+        for (Map.Entry<String, Writeable.Reader<? extends InternalAggregation>> bucketReaders : spec.bucketReaders.entrySet()) {
+            namedWriteableRegistry.register(InternalAggregation.class, bucketReaders.getKey(), bucketReaders.getValue());
+        }
+    }
+
+    public static class PipelineAggregationSpec {
```

Comment:
I built this to mimic `AggregationSpec`. I'm not sure that we need the flexibility it offers though. Do you think this is a good idea or should I make result readers required and bucket readers optional?